### PR TITLE
generate reports about the contributors

### PR DESCRIPTION
**Pull request by nikomatsakis**
Big overhaul of the setup. We still use the same metrics, but we now collect them into digested form. This code is a bit of a mess because I glued together some stuff I wrote, but it seems to work. 

Some notes for future:

* I don't think the intermediate "inputs" directory adds any value; we should probably remove it and instead have producers just produce typed structs instead of vectors of strings.
* The "top contributors" and "top crates" code in the reports should *really* be a producer itself, but I was too lazy to do that rewrite. It should probably also not be written synchronously?

### CODE REVIEW

File: book/src/report.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Generating a report
+
+Optopodi is geared to generate **reports**. These are CSV files that summarize important information about your repository or repositories. To create a report you need to take the following steps:
+
+- Make a directory `$DIR` for the report. We recommend a name like `data/2021-06-26`.
+- Create a `report.toml` file in that directory. You can start with [the template](https://github.com/optopodi/optopodi/blob/main/report-template.toml) and customize it.
```
Comment (chazkiker2) [inline book/src/report.md:6]: broken link for template at https://github.com/optopodi/optopodi/blob/main/report-template.toml

Comment (chazkiker2) [reply book/src/report.md:6]: ah maybe this won't be broken after merge... I see. You're two steps ahead 😉

File: src/report/repo_info.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+use super::{repo_participant::RepoParticipant, Report, ReportConfig};
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct RepoInfos {
+    pub repos: HashMap<String, RepoInfo>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct RepoInfo {
+    #[serde(rename = "#")]
```
Comment (chazkiker2) [inline src/report/repo_info.rs:17]: smart to use `serde(rename)` for column names. This is a much cleaner approach.

**Review by chazkiker2 — APPROVED** [review]
Hats off, @nikomatsakis. 

You mentioned lots of cleanups, and I'm sure there's plenty of refinements this code will undergo. I'll have to spend more time with it to really nail down where improvements could pop up. 

But I would be happy for this ginormous PR to get merged in and we can split out further refactorings into more granular Issues/PRs?

Up to you. But I approve. 

I definitely think heading in this direction is a **--much--** stronger path

File: book/src/report.md (hunk 1 above)
Comment (nikomatsakis) [reply book/src/report.md:6]: yes =)

**Comment by nikomatsakis** [comment]
I definitely want to merge as is and do follow-ups. Going to merge!

**Pull request merged**